volatility3/framework/plugins/windows/writecrashdump.py

```diff
@@ -0,0 +1,140 @@
+
+        number_processors = kernel.object("unsigned int", offset=kernel.get_symbol("KeNumberProcessors").address)
+        dump_header.NumberProcessors.write(number_processors)
+
+        dump_header.DirectoryTableBase.write(primary.config['page_map_offset'])
+        if dump_header.has_member("PaeEnabled"):
+            dump_header.PaeEnabled.write(int(is_pae))
+
+        dump_header.PfnDataBase.write(kdbg.MmPfnDatabase)
+        dump_header.PsLoadedModuleList.write(kdbg.PsLoadedModuleList)
+        dump_header.PsActiveProcessHead.write(kdbg.PsActiveProcessHead)
+        dump_header.DumpType.write(1)
+
+        dump_header.SystemTime.write(kuser.SystemTime.cast('unsigned long long'))
+
+        dump_header.BugCheckCode.write(0)
+        dump_header.BugCheckCodeParameter.write([0, 0, 0, 0])
+
+        blank_len = dump_header.Exception.vol.offset - dump_header.ContextRecord.vol.offset
+        header_layer.write(dump_header.ContextRecord.vol.offset, b"\x00" * blank_len)
+
+        dump_header.Comment.write(
+            bytes("Volatility 3 {} generated crashdump file\x00".format(constants.PACKAGE_VERSION),
+                  'latin-1'))
+
+        # Write the actual data
+        virtual_layer = context.layers[layer_name]
+        physical_layer_name = cls.get_physical_layer_name(context, virtual_layer)
+        physical_layer = context.layers[physical_layer_name]
+
+        page_count = (physical_layer.maximum_address - physical_layer.minimum_address) // 0x1000
+        dump_header.PhysicalMemoryBlockBuffer.NumberOfRuns.write(1)
+        dump_header.PhysicalMemoryBlockBuffer.NumberOfPages.write(page_count)
+        run0 = dump_header.PhysicalMemoryBlockBuffer.Run[0]
+        run0.BasePage.write(physical_layer.minimum_address)
+        run0.PageCount.write(page_count)
+
+        dump_header.RequiredDumpSpace.write((page_count + 2) * 0x1000)
+
+        # We don't try any form of compression, but just write the data as one large run
+        with open_method('crash.dmp') as f:
+            # We want to include the maxmium address
+            header_data = header_layer.read(0, header_layer.maximum_address + 1)
+            f.write(header_data)
+            for offset in range(physical_layer.minimum_address, physical_layer.maximum_address + 1, 0x1000):
+                f.write(physical_layer.read(offset, 0x1000, pad = True))
+
+            # Fix KDBG in the dump if it was encoded
+            decoded_data = context.layers[kdbg.vol.layer_name].read(kdbg.vol.offset, kdbg.size())
+            kdbg_physical_address = primary.translate(kdbg.vol.offset)[0]
+            kdbg_file_location = (header_layer.maximum_address + 1) + kdbg_physical_address - physical_layer.minimum_address
+            f.seek(kdbg_file_location)
+            f.write(decoded_data)
+
+    def _generator(self):
+        self.write_crashdump(self.context, self.config['primary'], self.config['nt_symbols'], self._file_handler)
+        yield 0, ('Done',)
+
+    def run(self):
```

Comment:
I think it'd be nice to have a Return Type here, too!

Reply:
I have not actually written that code. It is a copy-paste code from @ikelos.
Either way I have checked other plugins and they also many of them don't seem to have a return type for the `_generator` or `_run` functions.

Reply:
It's possible like this. (I like all your work, but This comment is intended to utilize the Github Suggestion function.)

```suggestion
    def run(self) -> renderers.TreeGrid:
```

Reply:
Yeah, we haven't historically, and it won't hold up merging, but if you're happy to have it, so are we.  5:)

Reply:
I have added the return type for the `run` function